Source/Gauntlet/Assertion/Assertion.swift

```diff
@@ -0,0 +1,251 @@
+//
+//  Assertion.swift
+//
+//  MIT License
+//
+//  Copyright (c) [2023] The Kroger Co. All rights reserved.
+//
+//  Permission is hereby granted, free of charge, to any person obtaining a copy
+//  of this software and associated documentation files (the "Software"), to deal
+//  in the Software without restriction, including without limitation the rights
+//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+//  copies of the Software, and to permit persons to whom the Software is
+//  furnished to do so, subject to the following conditions:
+//
+//  The above copyright notice and this permission notice shall be included in all
+//  copies or substantial portions of the Software.
+//
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+//  SOFTWARE.
+
+import Foundation
+
+/// An assertion that has been made on a value.
+///
+/// This is the main type in `Gauntlet`, and the type on which you will make assertions. An `Asertion` instance
+/// can be created by calling the `Assert(that: ...)` instance function on `XCTestCase`.
+///
+/// `Assertion`s of a specific type are extended to provide assertion functions relevant to that type. These functions return
+/// a new assertion allowing multiple assertions to be chained.
+///
+/// ```
+/// let boolValue: Bool = ...
+///
+/// Assert(that: boolValue).isTrue()
+///
+/// let result: Result<String, Error> = ...
+/// Assert(that: result)
+///     .isSuccess()
+///     .isEqualTo("Expected String")
+///
+/// let failureResult: Result<String, Error> = ...
+/// Assert(that: failureResult)
+///     .isFailure()
+/// ```
+/// 
+/// `Assertion` can be extended further to easily create new assertion functions that suit your needs or types.
+public final class Assertion<Value> {
+
+    /// The result of the assertion.
+    let result: AssertionResult<Value>
+
+    /// The name of the assertion. This should describe what the assertion is validating.
+    let name: String
+
+    /// The path to the source file in which the assertion exists.
+    let filePath: String
+
+    /// The line number of the assertion. This can be captured at the call site with the `#line` default parameter.
+    let lineNumber: Int
+
+    /// The `FailureRecorder` instance that failures should be recorded to.
+    let recorder: FailureRecorder
+
+    /// Indicates whether this assertion is the root assertion or if it spawned off of the root.
+    let isRoot: Bool
```

Comment:
I also converted all of these properties to internal. They were public initially, but I don't know if there's a good reason for any of them to be public. I kind of like the idea of the only public code on `Assertion` being the operators and the two `evaluate` functions that operators call. It also feels like it clarifies the API when implementing an operator as the only thing you can do is call `evaluate`.